apps/gdalalg_vector_reproject.cpp

```diff
@@ -0,0 +1,149 @@
+    {
+        m_layers.push_back(std::move(poLayer));
+    }
+
+    int GetLayerCount() override
+    {
+        return static_cast<int>(m_layers.size());
+    }
+
+    OGRLayer *GetLayer(int idx) override
+    {
+        return idx >= 0 && idx < GetLayerCount() ? m_layers[idx].get()
+                                                 : nullptr;
+    }
+};
+}  // namespace
+
+/************************************************************************/
+/*            GDALVectorReprojectAlgorithm::RunStep()                   */
+/************************************************************************/
+
+bool GDALVectorReprojectAlgorithm::RunStep(GDALProgressFunc, void *)
+{
+    CPLAssert(m_inputDataset.GetDatasetRef());
+    CPLAssert(m_outputDataset.GetName().empty());
+    CPLAssert(!m_outputDataset.GetDatasetRef());
+
+    std::unique_ptr<OGRSpatialReference> poSrcCRS;
+    if (!m_srsCrs.empty())
+    {
+        poSrcCRS = std::make_unique<OGRSpatialReference>();
+        if (poSrcCRS->SetFromUserInput(m_srsCrs.c_str()) != OGRERR_NONE)
+        {
+            ReportError(CE_Failure, CPLE_AppDefined,
+                        "Invalid value for '--src-crs'");
+            return false;
+        }
+        poSrcCRS->SetAxisMappingStrategy(OAMS_TRADITIONAL_GIS_ORDER);
+    }
+
+    OGRSpatialReference oDstCRS;
+    if (oDstCRS.SetFromUserInput(m_dstCrs.c_str()) != OGRERR_NONE)
+    {
+        ReportError(CE_Failure, CPLE_AppDefined,
+                    "Invalid value for '--dst-crs'");
+        return false;
+    }
+    oDstCRS.SetAxisMappingStrategy(OAMS_TRADITIONAL_GIS_ORDER);
+
+    auto poSrcDS = m_inputDataset.GetDatasetRef();
+
+    auto reprojectedDataset =
+        std::make_unique<GDALVectorReprojectAlgorithmDataset>();
+    reprojectedDataset->SetDescription(poSrcDS->GetDescription());
+
+    const int nLayerCount = poSrcDS->GetLayerCount();
+    bool ret = true;
+    for (int i = 0; i < nLayerCount; ++i)
+    {
+        auto poSrcLayer = poSrcDS->GetLayer(i);
+        ret = ret && (poSrcLayer != nullptr);
+        if (poSrcLayer)
+        {
+            OGRSpatialReference *poSrcLayerCRS;
+            if (poSrcCRS)
+                poSrcLayerCRS = poSrcCRS.get();
+            else
+                poSrcLayerCRS = poSrcLayer->GetSpatialRef();
+            if (!poSrcLayerCRS)
+            {
+                ReportError(CE_Failure, CPLE_AppDefined,
+                            "Layer '%s' has no spatial reference system",
+                            poSrcLayer->GetName());
+                return false;
+            }
+            auto poCT = std::unique_ptr<OGRCoordinateTransformation>(
+                OGRCreateCoordinateTransformation(poSrcLayerCRS, &oDstCRS));
+            auto poReversedCT = std::unique_ptr<OGRCoordinateTransformation>(
+                OGRCreateCoordinateTransformation(&oDstCRS, poSrcLayerCRS));
+            ret = ret && (poCT != nullptr) && (poReversedCT != nullptr);
```

Comment:
I don't understand this logic. Once `ret` goes false, it can never go true. But you keep going with the algorithm loop? What's going on? Why not just exit? Seems to deserve at least a comment.

It's also odd to have the check for `poCT` and `poReversedCT` twice with two different syntaxes.

Reply:
I've reworked the code to be more straightforward:

```
bool ret = true;
    for (int i = 0; ret && i < nLayerCount; ++i)
    {
        auto poSrcLayer = poSrcDS->GetLayer(i);
        ret = (poSrcLayer != nullptr);
        if (ret)
        {
            OGRSpatialReference *poSrcLayerCRS;
            if (poSrcCRS)
                poSrcLayerCRS = poSrcCRS.get();
            else
                poSrcLayerCRS = poSrcLayer->GetSpatialRef();
            if (!poSrcLayerCRS)
            {
                ReportError(CE_Failure, CPLE_AppDefined,
                            "Layer '%s' has no spatial reference system",
                            poSrcLayer->GetName());
                return false;
            }
            auto poCT = std::unique_ptr<OGRCoordinateTransformation>(
                OGRCreateCoordinateTransformation(poSrcLayerCRS, &oDstCRS));
            auto poReversedCT = std::unique_ptr<OGRCoordinateTransformation>(
                OGRCreateCoordinateTransformation(&oDstCRS, poSrcLayerCRS));
            ret = (poCT != nullptr) && (poReversedCT != nullptr);
            if (ret)
            {
                reprojectedDataset->AddLayer(std::make_unique<OGRWarpedLayer>(
                    poSrcLayer, /* iGeomField = */ 0,
                    /*bTakeOwnership = */ false, poCT.release(),
                    poReversedCT.release()));
            }
        }
    }
```
We could indeed use break / early return. The current way helps gaming a bit line code coverage...